test/backend/variables.py

```diff
@@ -42,6 +42,7 @@ def get_args_from_env():
     TEST_DYNAMIC = os.getenv("TEST_DYNAMIC")
     TEST_CONSTANT = os.getenv("TEST_CONSTANT")
     TEST_SIGNATURE = os.getenv("TEST_SIGNATURE")
```

Comment:
Not particular for this PR: can we use argument to the test.py instead of env variable for these options? I assume that options are better than env variable for such purpose.

Reply:
Not necessarily. Depending on the situation, options can be tricky to set due to the command string being interpreted multiple times, as you can see from the following example:
```
if (${PYTEST_XDIST_FOUND} EQUAL 0)
  message(STATUS "Parallel backend tests   : ON")
  set(BACKEND_TEST_COMMAND "${Python3_EXECUTABLE}" "-m" "pytest")
  set(BACKEND_TEST_ARGS "--forked" "-n" "$$\{NPROC:-auto\}" "-q" "--silent")
```
Here you need two `$` because cmake interprets the string first and makes that a single `$` for bash to consume. Similarly for the escaped `\{` so that after the cmake pass you are left with the `{` for bash to consume. For situations like this, it's typically a lot simpler to use env vars to set the parameters.

There are also situations where you definitely want to use env vars. For example, there might be a chain of invocation, i.e., script A invokes B, which invokes C. If you want to pass something to C via command line argument, you will have to write code in A and B to pass on the option. But with env var, you just need to set it when invoking A and it gets automatically propagated to all the subprocesses (B and C).

So in general, often used options should be settable by both command line argument and env var.

Reply:
@gongsu832 thanks for the comments.